HaloMD/master_server/src/main.rs

```diff
@@ -0,0 +1,205 @@
+                match File::open(BLACKLIST_FILE) {
+                    Ok(file) => {
+                        let reader = BufReader::new(&file);
+                        match reader.lines().collect() {
+                            Err(_) => {
+                                blacklist_ref.clear();
+                            },
+                            Ok(t) => {
+                                *blacklist_ref = t;
+                            }
+                        }
+                    },
+                    Err(_) => {}
+                };
+            }
+            thread::sleep_ms(BLACKLIST_UPDATE_TIME * 1000);
+        }
+    });
+
+    // TCP server thread. This is for the HaloMD application.
+    thread::spawn(move || {
+        loop {
+            for stream in client_socket.incoming() {
+                let mut client = match stream {
+                    Err(_) => continue,
+                    Ok(the_stream) => the_stream
+                };
+                // Unwrap the IP.
+                let ip = match client.peer_addr() {
+                    Err(_) => continue,
+                    Ok(ip) => ip.to_string().split(":").next().unwrap().to_string()
+                };
+
+                let mut ips = String::new();
+
+                let servers_ref = servers_mut_tcp.lock().unwrap();
+                let servers = (*servers_ref).iter();
+
+                for j in servers {
+                    ips = ips + &(j.to_string()) + "\n";
+                }
+
+                // Some number placed after the requester's IP. If you ask me, the source code was abducted by aliens, and this is a tracking number. Regardless, it's needed.
+                ips = ips + &ip + ":49149:3425";
+                let _ = client.write_all(ips.as_bytes());
+            }
+        }
+    });
+
+
+
+    // UDP server is run on the main thread. Servers broadcast their presence here.
+
+    // These are the allowed game versions. HaloMD and Halo PC 1.09 uses 01.00.09.0620, while Halo PC servers on 1.10 use 01.00.10.0621 (these are also joinable).
+    let game_versions = [ "01.00.09.0620".to_string(), "01.00.10.0621".to_string() ];
+
+    let mut buffer = [0; 2048];
+    loop {
+        let (length, source) = match halo_socket.recv_from(&mut buffer) {
+            Err(_) => continue,
+            Ok((length, source)) => (length,source)
+        };
+        if length == 0 {
+            continue;
+        }
+        let client_ip = source.to_string().split(":").next().unwrap().to_string();
+
+        if buffer[0] == 3 || buffer[0] == 8 {
+            let blacklist_ref = blacklist_udp.lock().unwrap();
+            if blacklist_ref.contains(&client_ip) {
+                continue;
+            }
+
+            // Heartbeat packet.
+            if buffer[0] == 3 {
+
+                let mut servers = servers_mut_udp.lock().unwrap();
+
+                let packet = HeartbeatPacket::from_buffer(&buffer, length);
```

Comment:
First thing is first. We want to use the `len()` on the slice we pass to `from_buffer`. Get the sub-slice from the array by passing `&buffer[..length]`